### PR TITLE
refactor: harden tests by adjusting current array-providing fixtures

### PR DESCRIPTION
**Pull request by douglasdavis**
This PR is an experiment in hardening our existing tests.

Many of our tests are performed by using the `daa` and `caa` fixtures.

- `daa` is a dask-awkward `Array` that is created via `dak.from_json`. 
- `caa` is a concrete awkward `Array` created from `ak.from_json` followed by `ak.concatenate`. 

the structure of the array is:

```
n * {
    points: var * {
        x: int64,
        y: int64
    }
}
```
(a record array with a single entry (`points`) which each contain a variable number of `(x, y)` pairs)

This makes it easy to write a test like:

```python
def test_sum_axis1(daa, caa):
    assert_eq(
        dak.sum(daa.points.x, axis=1),
        dak.sum(caa.points.x, axis=1),
    )
```

Currently dask-awkward JSON reads do not support the `project_columns` interface (though it is something on our todo list).

So `daa.points.x` will not be optimized to read only the `points.x` field.

In this PR I've made the `daa` fixture originate via a `dak.from_parquet` call.

### CODE REVIEW

**Comment by douglasdavis** [comment]
Something that's popping up in a number of tests: the `_unproject_layout` function's final `else` case raises due to an unexpected combination. We're getting a case where the form is `None` and the layout is a `RecordArray` content.

```
AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
```

a test to trigger this (not the only one):

```python
$ pytest tests/test_inspect.py::test_sample
```

full list of failures due to that assertion:

```
FAILED tests/test_core.py::test_compute_typetracer - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_core.py::test_scalar_to_delayed[False] - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_core.py::test_optimize_chain_single - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_getitem.py::test_single_int - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_getitem.py::test_single_partition - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_inspect.py::test_sample - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_io.py::test_to_and_from_delayed[False] - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_io.py::test_to_delayed[False] - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_io.py::test_to_bag - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_io.py::test_to_dataframe[False] - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
FAILED tests/test_structure.py::test_full_like[datetime64] - AssertionError: unexpected combination: <class 'NoneType'> and <class 'awkward.contents.recordarray.RecordArray'>
```

simple repro:

```python
import dask_awkward as dak
import awkward as ak

arr = ak.Array(
    [
        {"x": [1,2,3], "y": 0},
        {"x": [4,5], "y": 1},
    ]
)
ak.to_parquet(arr, "test.parquet")
darr = dak.from_parquet("test.parquet")
samp = dak.sample(darr, probability=0.8)
samp.compute()
```
I'll see if I can track it down some, but tagging @agoose77 for visibility

**Comment by agoose77** [comment]
Thanks @douglasdavis. `unproject_layout` always needs a `Form`, so I would expect the caller to ensure that it passes in a valid `Form` object before calling the function?

**Comment by douglasdavis** [comment]
thanks @agoose77 that was quick! Yes I tracked down the issue. I was thinking we perhaps hit a bad combo in the recursion, but I was wrong. We are currently explicitly passing  a `None` form at an explicit call site in cases _when the column optimization fails_. Just pushed a simple fix to bypass running _unproject_columns when the column opt fails.

File: pyproject.toml [hunk 1]
```diff
@@ -27,6 +27,7 @@ classifiers = [
   "Topic :: Scientific/Engineering",
 ]
 dependencies = [
+  "numpy <=1.24.3",
```
Comment (agoose77) [inline pyproject.toml:30]: I'd suggest moving this to the CI resolution, if required; we'll patch this upstream in short order, and then make a new release!

File: src/dask_awkward/lib/core.py [hunk 2]
```diff
@@ -2051,3 +2059,13 @@ def typetracer_from_form(form: Form) -> ak.Array:
...
+
+
+def make_unknown_length(array: ak.Array) -> ak.Array:
+    return ak.Array(
+        ak.to_backend(
+            array,
+            "typetracer",
+            highlevel=False,
+        ).to_typetracer(forget_length=True)
+    )
```
Comment (agoose77) [inline src/dask_awkward/lib/core.py:2071]: This can just be `to_layout(array).to_typetracer(...)` :)

File: pyproject.toml (hunk 1 above)
Comment (douglasdavis) [reply pyproject.toml]: right, I did this as a temporary sanity check just to see if GH actions would pass :) - gone now!

File: tests/test_behavior.py [hunk 3]
```diff
@@ -31,6 +31,7 @@ def non_dask_method(self, _dask_array_=None):
         return _dask_array_
 
 
+@pytest.mark.xfail(reason="NumPy 1.25 mixin __slots__ change")
```
Comment (agoose77) [inline tests/test_behavior.py:34]: We could use a conditional fail (check `awkward>XXX || numpy>=YYY`), or just leave it until we fix it!

Comment (douglasdavis) [reply tests/test_behavior.py]: thanks for the suggestion!

File: src/dask_awkward/lib/core.py (hunk 2 above)
Comment (douglasdavis) [reply src/dask_awkward/lib/core.py]: :+1:

**Pull request merged**